[TwigBridge] Change return type of Symfony\Bridge\Twig\AppVariable::getSession()

Dirkhuethorst opened the pull request:
Change the return type of AppVariable.php getSession from 'Session' to 'SessionInterface', to be in line with the return type of getSession in http-foundation/Request.php

| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #50789
| License       | MIT
| Doc PR        |

Dirkhuethorst committed:
[TwigBridge] Fix for GitHub ticket 50789

Change the return type of AppVariable.php getSession from 'Session' to 'SessionInterface', to be in line with the return type of getSession in http-foundation/Request.php

Dirkhuethorst committed:
[TwigBridge] Fix for GitHub ticket 50789

Change the return type of AppVariable.php getSession from 'Session' to 'FlashBagAwareSessionInterface'

GromNaN commented on src/Symfony/Bridge/Twig/AppVariable.php:
```diff
@@ -93,7 +93,7 @@ public function getRequest(): ?Request
      * Returns the current session.
      */
-    public function getSession(): ?Session
+    public function getSession(): ?FlashBagAwareSessionInterface
```
The return type could be the intersection of all the interfaces implemented by `Session` in `symfony/http-foundation: 5.4`.

```suggestion
    public function getSession(): ?SessionInterface&\Traversable&\Countable
```

derrabus replied:
@GromNaN Native intersection types are not nullable. And unions of a intersection type with the `null` type require PHP >= 8.2. That aside, your proposal generates the same problem as the current type: The HttpFoundation `Request` cannot guarantee that the session it returns is `Traversable` and/or `Countable`.

GromNaN replied:
Well, I've forgotten nullable in my proposal. Clearly, the return type cannot be changed in a totally backwards-compatible way.

Noticed [your proposition](https://github.com/symfony/symfony/pull/50794#issuecomment-1611409156) looks more correct.

Nyholm commented:
I see you replaced the return type from class `Session` to interface `FlashBagAwareSessionInterface`. This is not always correct. It could be a Session that do not implement FlashBag feature. 

I suggest to change it to `SessionInterface`

GromNaN reviewed (changes requested):
The `FlashBagAwareSessionInterface` was introduced in `symfony/http-foundation: 6.2` by #46491, and `symfony/twig-bridge: 6.2` allows `symfony/http-foundation: 5.4` in which the interface doesn't exit.
We usually don't update the dependencies constraints in maintenance branches except for security reasons. I don't think we should accept this change as a bugfix, it looks more like a new feature that would target the next version (6.4 ATM).

Also, `Symfony\Component\HttpFoundation\Session\Session` implements `\IteratorAggregate` and `\Countable` on top of `FlashBagAwareSessionInterface` or `SessionInterface`. Changing the return type to this single interface would reduce the functions that could be used in the templates.

Nyholm commented:
I was hesitating if this was a bugfix or not. But with @GromNaN arguments, I agree. This is a feature.

derrabus commented:
I think, if we change the return type to `?SessionInterface`, we can file it as a bugfix. Using a custom implementation of the `SessionInterface` should've worked in Symfony 5 and it stopped working because of a too narrow native return type in Symfony 6. I believe this breaking change was unintended.

Dirkhuethorst committed:
[TwigBridge] Fix for GitHub ticket 50789

Change the return type of AppVariable.php getSession from 'Session' to 'SessionInterface'

Dirkhuethorst commented:
I've changed the return type to SessionInterface, but now the Psalm check fails because in the AppVariable class the Flashbag of the session is accessed. 

`src/Symfony/Bridge/Twig/AppVariable.php:147:30: UndefinedInterfaceMethod: Method Symfony\Component\HttpFoundation\Session\SessionInterface::getFlashBag does not exist (see https://psalm.dev/181)`

I am not sure how to proceed with this issue

GromNaN commented:
You can `return []` if `$session` is not an instance of `FlashBagAwareSessionInterface`.

Dirkhuethorst commented:
> You can `return []` if `$session` is not an instance of `FlashBagAwareSessionInterface`.

Won't that break backwards compatibility  with http-foundation 5.4 where the FlashBagAwareSessionInterface does not exist?

Dirkhuethorst committed:
[TwigBridge] Fix for GitHub ticket 50789

Change the return type of AppVariable.php getSession from 'Session' to 'SessionInterface'

GromNaN commented on src/Symfony/Bridge/Twig/AppVariable.php:
```diff
@@ -143,6 +143,10 @@ public function getFlashes(string|array $types = null): array
             return [];
         }
 
+        if (!method_exists($session, 'getFlashBag')) {
```
You can add a comment that this can be updated in 7.0 to check `FlashBagAwareSessionInterface`, when `symfony/http-foundation: 6.4` is required.

Dirkhuethorst committed:
[TwigBridge] Fix for GitHub ticket 50789

Change the return type of AppVariable.php getSession from 'Session' to 'SessionInterface'

fabpot commented:
Thank you @Dirkhuethorst.

fabpot commented:
Merged, but Github didn't get the memo.
